### ISSUE TITLE
PEP 702: Missing warnings for implicit calls to dunders

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#6412]
PEP 702 says that any syntax that implicitly calls a deprecated function should produce a deprecation warning. Pyright doesn't seem to do this (https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigBMsEQsBjAQ3zoCguWAbNoUJQA8nToAuLlBlQAAgyasOWOgAoARHTBQUYWgFdCWKAD8NASmmyGwKBgyoAblhoO1xvsAtQAtAD4oQhgQCSgAOkjrGQVGZnZOTW1dfSgjEwoEPiQWJFoAIzAwPiw2FCgWUhcQYlJLaPosOwdC4vdPbz9A1r4wyPCG2KUE1SSdPVo2cXrZRuaMKbp2rC8AGih9AAtXMLB8gCtWGB8AqGA%2BMA4%2BqK5mFzY%2BDHgELDVTMXULKzvSx%2BfXj5qCzhBzOVwwdxfW5Ye5-RAA8RAqAAaigAEYrFwkHYJqJERYpLMEIJCFwgA):

```python
from typing_extensions import deprecated

class Odd:
    @deprecated("do not use ~")
    def __invert__(self) -> str: ...
    @deprecated("do not use implicit boolean conversion")
    def __bool__(self) -> bool: ...
    @deprecated("do not add")
    def __add__(self, other: object) -> float: ...

reveal_type(~Odd())  # str, no deprecation error
reveal_type(Odd().__invert__())  # deprecation error (as expected)
reveal_type(Odd() + 1)  # float, no deprecation error

if not Odd():  # no deprecation error
    pass
```

The `__invert__` and `__bool__` examples come from python/typeshed#11009 (`bool.__invert__` and `xml.etree.ElementTree.Element.__bool__` are deprecated); `__add__` is an explicit example in PEP 702.

**Comment by erictraut** [comment]
This is getting really deep into how the language is modeled and would be very difficult to do throughout the code logic. I'm going to reject this for now. I might reconsider this in the future for select operations if there is sufficient demand.

**Comment by erictraut** [comment]
I've added deprecated support for unary, binary, and augmented assignment operators. No pyright users have requested this functionality, but it is now covered by the conformance test suite so I feel compelled to add the support. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
